### ISSUE TITLE
Conversion to object must be explicit

### ISSUE THREAD

**Issue by jl0pd** [LanguageDev/Fresh-Language-suggestions#68]
`System.Object` is a root of type hierarchy in dotnet and all mainstream languages (C#, F#, VB) allows implicit upcasting to parent types. I agree with implicit conversion to base type, but `object` should be exception, that won't be allowed to implicitly converted.

There are some examples 

* Some APIs are expecting objects as inputs. `ReferenceEquals` or NUnit's `Assert.AreEqual` may behave unexpectedly for some developers. 

```csharp
Object.ReferenceEquals(1, 1) // false
Object.ReferenceEquals("q", 1) // false. This should not have compiled, but it's allowed
```

* You may want to return object is some cases from method, but cache value for performance reason

```csharp
object PerformComputation(int a, int b)
{
    return a + b; // implicit boxing, always produces new object
}

object PerformComputation(int a, int b)
{
    return GetBoxed(a + b); // boxing is explicit, potentially returning cached value
}
```

* Overload resolution may pick unexpected method

```csharp
void M(float a) { }
void M(object a) { }
void Foo() => M(5.0); // calls M(object)
```

---

Real world example: WPF's [`UncommonField<T>.SetValue`](https://github.com/dotnet/wpf/blob/v6.0.7/src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/UncommonField.cs#L54) boxes both input value and `_defaultValue` on each call, while it could perform better. It was [fixed](https://github.com/dotnet/wpf/blob/277533ff5586be9004251778aae6efd479c9d25d/src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/UncommonField.cs#L54) only in net7 timeline

**Comment by jl0pd** [comment]
Implicit upcasting of reference types should be allowed, since it's a basis of polymorphism. Requirement of casting to base type may quickly become annoying and will neglect value of this proposal, because people will just always think "alright, insert casting *again*". So I think that conversion of reference types to it's parent or interface still should be implicit, with exception of `object`. Not sure about conversion of struct to interface

**Comment by jl0pd** [comment]
Yeah, it's still dangerous, even with reference types:
```csharp
Client client = ...;
Order order = makeOrder(client);
client.Equals(order);
```

This will compile, even though it doesn't make any sense to do such operation (comparing models of different types)

**Comment by svick** [comment]
This is a weird idea, but maybe the language could hide `object.Equals(object)` on any type that is not `object` itself?

**Comment by LPeter1997** [comment]
Yeah this is why I semi-proposed explicit boxing instead, as that seems like a more logical addition to me. But other opinions would also be nice on this.

**Comment by WalkerCodeRanger** [comment]
I think that implicit boxing and the bad `Equals` API are the root of the problem. However, getting rid of implicit boxing causes a lot of problems around generics. If you have `where T : ISomething` then you assume you can pass `T` to a method expecting `ISomething` which only works because of implicit boxing. The same thing can happen on the `this` reference for traits or implicit interface implementation. I don't know what I would recommend.

**Comment by jl0pd** [comment]
This proposal is only for implicit cast to `System.Object`. Other forms of boxing is permitted

**Comment by Kuinox** [comment]
I'm not against having to explicitly box things.  
Boxing is a big source of accidental allocation that may not be even necessary.

**Comment by jl0pd** [comment]
CLR is smart enough to avoid allocations when it's unnecessary. Classes are never boxed, structs are not boxed when used in generics. Requiring explicit cast to interface will bring much hassle without benefits
